### PR TITLE
Remove excessive indirection with sending remote drawing commands (part 1)

### PR DESCRIPTION
**Pull request by kkinnunen-apple**
#### 5910db50c2cdffce5ef0fe0cd12f5925029d0ec0
<pre>
Remove excessive indirection with sending remote drawing commands (part 1)
<a href="https://bugs.webkit.org/show_bug.cgi?id=279759">https://bugs.webkit.org/show_bug.cgi?id=279759</a>
<a href="https://rdar.apple.com/136070541">rdar://136070541</a>

Reviewed by Simon Fraser.

DisplayList::Recorder interface would cause overly many virtual function
calls and would prevent backend-specific caching behavior.

Consider callgraph for `context-&gt;save()`:
Recorder::save() &lt;vfunc&gt;
 RemoteDisplayListRecorderProxy::recordSave() &lt;vfunc&gt;

Instead, fix it in this patch to be:
RemoteDisplayListRecorderProxy::recordSave() &lt;vfunc&gt;
  Recorder::updateStateForSave() &lt;normal func&gt;

In future patches, this allows useResource calls to record and return
backend command specific data without the need for this data to leak
into the WebCore. For example, current identifiers should not be defined
in WebCore at all, but they are, since useResource is there.

In this part 1, change the commands that are the most straight-forward.

* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawLinesForText::DrawLinesForText):
(WebCore::DisplayList::DrawLinesForText::apply const):
(WebCore::DisplayList::DrawLinesForText::dump const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::DrawLinesForText::point const):
(WebCore::DisplayList::DrawLinesForText::setBlockLocation): Deleted.
(WebCore::DisplayList::DrawLinesForText::blockLocation const): Deleted.
(WebCore::DisplayList::DrawLinesForText::localAnchor const): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::updateStateForSave):
(WebCore::DisplayList::Recorder::updateStateForRestore):
(WebCore::DisplayList::Recorder::updateStateForTranslate):
(WebCore::DisplayList::Recorder::updateStateForRotate):
(WebCore::DisplayList::Recorder::updateStateForScale):
(WebCore::DisplayList::Recorder::updateStateForConcatCTM):
(WebCore::DisplayList::Recorder::updateStateForSetCTM):
(WebCore::DisplayList::Recorder::updateStateForBeginTransparencyLayer):
(WebCore::DisplayList::Recorder::updateStateForEndTransparencyLayer):
(WebCore::DisplayList::Recorder::updateStateForResetClip):
(WebCore::DisplayList::Recorder::updateStateForClip):
(WebCore::DisplayList::Recorder::updateStateForClipRoundedRect):
(WebCore::DisplayList::Recorder::updateStateForClipOut):
(WebCore::DisplayList::Recorder::updateStateForClipOutRoundedRect):
(WebCore::DisplayList::Recorder::updateStateForClipPath):
(WebCore::DisplayList::Recorder::updateStateForApplyDeviceScaleFactor):
(WebCore::DisplayList::Recorder::setLineCap): Deleted.
(WebCore::DisplayList::Recorder::setLineDash): Deleted.
(WebCore::DisplayList::Recorder::setLineJoin): Deleted.
(WebCore::DisplayList::Recorder::setMiterLimit): Deleted.
(WebCore::DisplayList::Recorder::save): Deleted.
(WebCore::DisplayList::Recorder::restore): Deleted.
(WebCore::DisplayList::Recorder::translate): Deleted.
(WebCore::DisplayList::Recorder::rotate): Deleted.
(WebCore::DisplayList::Recorder::scale): Deleted.
(WebCore::DisplayList::Recorder::concatCTM): Deleted.
(WebCore::DisplayList::Recorder::setCTM): Deleted.
(WebCore::DisplayList::Recorder::beginTransparencyLayer): Deleted.
(WebCore::DisplayList::Recorder::endTransparencyLayer): Deleted.
(WebCore::DisplayList::Recorder::drawRect): Deleted.
(WebCore::DisplayList::Recorder::drawLine): Deleted.
(WebCore::DisplayList::Recorder::drawLinesForText): Deleted.
(WebCore::DisplayList::Recorder::drawDotsForDocumentMarker): Deleted.
(WebCore::DisplayList::Recorder::drawEllipse): Deleted.
(WebCore::DisplayList::Recorder::drawPath): Deleted.
(WebCore::DisplayList::Recorder::drawFocusRing): Deleted.
(WebCore::DisplayList::Recorder::fillRect): Deleted.
(WebCore::DisplayList::Recorder::fillRoundedRect): Deleted.
(WebCore::DisplayList::Recorder::fillRectWithRoundedHole): Deleted.
(WebCore::DisplayList::Recorder::fillEllipse): Deleted.
(WebCore::DisplayList::Recorder::strokeRect): Deleted.
(WebCore::DisplayList::Recorder::strokeEllipse): Deleted.
(WebCore::DisplayList::Recorder::clearRect): Deleted.
(WebCore::DisplayList::Recorder::applyStrokePattern): Deleted.
(WebCore::DisplayList::Recorder::applyFillPattern): Deleted.
(WebCore::DisplayList::Recorder::drawControlPart): Deleted.
(WebCore::DisplayList::Recorder::resetClip): Deleted.
(WebCore::DisplayList::Recorder::clip): Deleted.
(WebCore::DisplayList::Recorder::clipRoundedRect): Deleted.
(WebCore::DisplayList::Recorder::clipOut): Deleted.
(WebCore::DisplayList::Recorder::clipOutRoundedRect): Deleted.
(WebCore::DisplayList::Recorder::clipPath): Deleted.
(WebCore::DisplayList::Recorder::drawVideoFrame): Deleted.
(WebCore::DisplayList::Recorder::applyDeviceScaleFactor): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
(WebCore::DisplayList::Recorder::recordDrawPattern):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::save):
(WebCore::DisplayList::RecorderImpl::restore):
(WebCore::DisplayList::RecorderImpl::translate):
(WebCore::DisplayList::RecorderImpl::rotate):
(WebCore::DisplayList::RecorderImpl::scale):
(WebCore::DisplayList::RecorderImpl::setCTM):
(WebCore::DisplayList::RecorderImpl::concatCTM):
(WebCore::DisplayList::RecorderImpl::setLineCap):
(WebCore::DisplayList::RecorderImpl::setLineDash):
(WebCore::DisplayList::RecorderImpl::setLineJoin):
(WebCore::DisplayList::RecorderImpl::setMiterLimit):
(WebCore::DisplayList::RecorderImpl::resetClip):
(WebCore::DisplayList::RecorderImpl::clip):
(WebCore::DisplayList::RecorderImpl::clipRoundedRect):
(WebCore::DisplayList::RecorderImpl::clipOut):
(WebCore::DisplayList::RecorderImpl::clipOutRoundedRect):
(WebCore::DisplayList::RecorderImpl::clipPath):
(WebCore::DisplayList::RecorderImpl::beginTransparencyLayer):
(WebCore::DisplayList::RecorderImpl::endTransparencyLayer):
(WebCore::DisplayList::RecorderImpl::drawRect):
(WebCore::DisplayList::RecorderImpl::drawLine):
(WebCore::DisplayList::RecorderImpl::drawLinesForText):
(WebCore::DisplayList::RecorderImpl::drawDotsForDocumentMarker):
(WebCore::DisplayList::RecorderImpl::drawEllipse):
(WebCore::DisplayList::RecorderImpl::drawPath):
(WebCore::DisplayList::RecorderImpl::drawFocusRing):
(WebCore::DisplayList::RecorderImpl::fillRect):
(WebCore::DisplayList::RecorderImpl::fillRoundedRect):
(WebCore::DisplayList::RecorderImpl::fillRectWithRoundedHole):
(WebCore::DisplayList::RecorderImpl::fillEllipse):
(WebCore::DisplayList::RecorderImpl::drawVideoFrame):
(WebCore::DisplayList::RecorderImpl::strokeRect):
(WebCore::DisplayList::RecorderImpl::strokeEllipse):
(WebCore::DisplayList::RecorderImpl::clearRect):
(WebCore::DisplayList::RecorderImpl::drawControlPart):
(WebCore::DisplayList::RecorderImpl::applyStrokePattern):
(WebCore::DisplayList::RecorderImpl::applyFillPattern):
(WebCore::DisplayList::RecorderImpl::applyDeviceScaleFactor):
(WebCore::DisplayList::RecorderImpl::recordSave): Deleted.
(WebCore::DisplayList::RecorderImpl::recordRestore): Deleted.
(WebCore::DisplayList::RecorderImpl::recordTranslate): Deleted.
(WebCore::DisplayList::RecorderImpl::recordRotate): Deleted.
(WebCore::DisplayList::RecorderImpl::recordScale): Deleted.
(WebCore::DisplayList::RecorderImpl::recordSetCTM): Deleted.
(WebCore::DisplayList::RecorderImpl::recordConcatenateCTM): Deleted.
(WebCore::DisplayList::RecorderImpl::recordSetLineCap): Deleted.
(WebCore::DisplayList::RecorderImpl::recordSetLineDash): Deleted.
(WebCore::DisplayList::RecorderImpl::recordSetLineJoin): Deleted.
(WebCore::DisplayList::RecorderImpl::recordSetMiterLimit): Deleted.
(WebCore::DisplayList::RecorderImpl::recordResetClip): Deleted.
(WebCore::DisplayList::RecorderImpl::recordClip): Deleted.
(WebCore::DisplayList::RecorderImpl::recordClipRoundedRect): Deleted.
(WebCore::DisplayList::RecorderImpl::recordClipOut): Deleted.
(WebCore::DisplayList::RecorderImpl::recordClipOutRoundedRect): Deleted.
(WebCore::DisplayList::RecorderImpl::recordClipOutToPath): Deleted.
(WebCore::DisplayList::RecorderImpl::recordClipPath): Deleted.
(WebCore::DisplayList::RecorderImpl::recordBeginTransparencyLayer): Deleted.
(WebCore::DisplayList::RecorderImpl::recordEndTransparencyLayer): Deleted.
(WebCore::DisplayList::RecorderImpl::recordDrawRect): Deleted.
(WebCore::DisplayList::RecorderImpl::recordDrawLine): Deleted.
(WebCore::DisplayList::RecorderImpl::recordDrawLinesForText): Deleted.
(WebCore::DisplayList::RecorderImpl::recordDrawDotsForDocumentMarker): Deleted.
(WebCore::DisplayList::RecorderImpl::recordDrawEllipse): Deleted.
(WebCore::DisplayList::RecorderImpl::recordDrawPath): Deleted.
(WebCore::DisplayList::RecorderImpl::recordDrawFocusRingPath): Deleted.
(WebCore::DisplayList::RecorderImpl::recordDrawFocusRingRects): Deleted.
(WebCore::DisplayList::RecorderImpl::recordFillRect): Deleted.
(WebCore::DisplayList::RecorderImpl::recordFillRectWithColor): Deleted.
(WebCore::DisplayList::RecorderImpl::recordFillRectWithGradient): Deleted.
(WebCore::DisplayList::RecorderImpl::recordFillRectWithGradientAndSpaceTransform): Deleted.
(WebCore::DisplayList::RecorderImpl::recordFillCompositedRect): Deleted.
(WebCore::DisplayList::RecorderImpl::recordFillRoundedRect): Deleted.
(WebCore::DisplayList::RecorderImpl::recordFillRectWithRoundedHole): Deleted.
(WebCore::DisplayList::RecorderImpl::recordFillEllipse): Deleted.
(WebCore::DisplayList::RecorderImpl::recordDrawVideoFrame): Deleted.
(WebCore::DisplayList::RecorderImpl::recordStrokeRect): Deleted.
(WebCore::DisplayList::RecorderImpl::recordStrokeEllipse): Deleted.
(WebCore::DisplayList::RecorderImpl::recordClearRect): Deleted.
(WebCore::DisplayList::RecorderImpl::recordDrawControlPart): Deleted.
(WebCore::DisplayList::RecorderImpl::recordApplyStrokePattern): Deleted.
(WebCore::DisplayList::RecorderImpl::recordApplyFillPattern): Deleted.
(WebCore::DisplayList::RecorderImpl::recordApplyDeviceScaleFactor): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
(WebCore::DisplayList::RecorderImpl::recordDrawPattern):
* Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::save):
(WebKit::RemoteDisplayListRecorderProxy::restore):
(WebKit::RemoteDisplayListRecorderProxy::translate):
(WebKit::RemoteDisplayListRecorderProxy::rotate):
(WebKit::RemoteDisplayListRecorderProxy::scale):
(WebKit::RemoteDisplayListRecorderProxy::setCTM):
(WebKit::RemoteDisplayListRecorderProxy::concatCTM):
(WebKit::RemoteDisplayListRecorderProxy::setLineCap):
(WebKit::RemoteDisplayListRecorderProxy::setLineDash):
(WebKit::RemoteDisplayListRecorderProxy::setLineJoin):
(WebKit::RemoteDisplayListRecorderProxy::setMiterLimit):
(WebKit::RemoteDisplayListRecorderProxy::clip):
(WebKit::RemoteDisplayListRecorderProxy::clipRoundedRect):
(WebKit::RemoteDisplayListRecorderProxy::clipOut):
(WebKit::RemoteDisplayListRecorderProxy::clipOutRoundedRect):
(WebKit::RemoteDisplayListRecorderProxy::clipPath):
(WebKit::RemoteDisplayListRecorderProxy::resetClip):
(WebKit::RemoteDisplayListRecorderProxy::beginTransparencyLayer):
(WebKit::RemoteDisplayListRecorderProxy::endTransparencyLayer):
(WebKit::RemoteDisplayListRecorderProxy::drawRect):
(WebKit::RemoteDisplayListRecorderProxy::drawLine):
(WebKit::RemoteDisplayListRecorderProxy::drawLinesForText):
(WebKit::RemoteDisplayListRecorderProxy::drawDotsForDocumentMarker):
(WebKit::RemoteDisplayListRecorderProxy::drawEllipse):
(WebKit::RemoteDisplayListRecorderProxy::drawPath):
(WebKit::RemoteDisplayListRecorderProxy::drawFocusRing):
(WebKit::RemoteDisplayListRecorderProxy::fillRect):
(WebKit::RemoteDisplayListRecorderProxy::fillRoundedRect):
(WebKit::RemoteDisplayListRecorderProxy::fillRectWithRoundedHole):
(WebKit::RemoteDisplayListRecorderProxy::fillEllipse):
(WebKit::RemoteDisplayListRecorderProxy::drawVideoFrame):
(WebKit::RemoteDisplayListRecorderProxy::strokeRect):
(WebKit::RemoteDisplayListRecorderProxy::strokeEllipse):
(WebKit::RemoteDisplayListRecorderProxy::clearRect):
(WebKit::RemoteDisplayListRecorderProxy::drawControlPart):
(WebKit::RemoteDisplayListRecorderProxy::applyStrokePattern):
(WebKit::RemoteDisplayListRecorderProxy::applyFillPattern):
(WebKit::RemoteDisplayListRecorderProxy::applyDeviceScaleFactor):
(WebKit::RemoteDisplayListRecorderProxy::recordSave): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordRestore): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordTranslate): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordRotate): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordScale): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordSetCTM): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordConcatenateCTM): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordSetLineCap): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordSetLineDash): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordSetLineJoin): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordSetMiterLimit): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordClip): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordClipRoundedRect): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordClipOut): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordClipOutRoundedRect): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordClipOutToPath): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordClipPath): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordResetClip): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordBeginTransparencyLayer): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordEndTransparencyLayer): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordDrawRect): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordDrawLine): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordDrawLinesForText): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordDrawDotsForDocumentMarker): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordDrawEllipse): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordDrawPath): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordDrawFocusRingPath): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordDrawFocusRingRects): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordFillRect): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordFillRectWithColor): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordFillRectWithGradient): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordFillRectWithGradientAndSpaceTransform): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordFillCompositedRect): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordFillRoundedRect): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordFillRectWithRoundedHole): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordFillEllipse): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordDrawVideoFrame): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeRect): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeEllipse): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordClearRect): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordDrawControlPart): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordApplyStrokePattern): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordApplyFillPattern): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordApplyDeviceScaleFactor): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
(WebKit::RemoteDisplayListRecorderProxy::recordDrawPattern):

Canonical link: <a href="https://commits.webkit.org/284963@main">https://commits.webkit.org/284963@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/2a06e405e630d45939804e9469b00fb6fa7ce7c1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69591 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48991 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22290 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73676 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20749 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71708 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56792 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20600 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55322 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13784 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72657 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44678 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60067 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35801 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41344 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17498 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19126 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63279 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17843 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75386 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13573 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17073 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62984 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13613 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60150 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62901 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15724 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10928 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4543 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44795 "Built successfully") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45869 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47064 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45610 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp [hunk 1]
```diff
@@ -451,9 +451,8 @@ void DrawLine::dump(TextStream& ts, OptionSet<AsTextFlag>) const
     ts.dumpProperty("point-2", point2());
 }
 
-DrawLinesForText::DrawLinesForText(const FloatPoint& blockLocation, const FloatSize& localAnchor, const DashArray& widths, float thickness, bool printing, bool doubleLines, StrokeStyle style)
-    : m_blockLocation(blockLocation)
-    , m_localAnchor(localAnchor)
+DrawLinesForText::DrawLinesForText(const FloatPoint& point, const DashArray& widths, float thickness, bool printing, bool doubleLines, StrokeStyle style)
+    : m_point(point)
```
Comment (shallawa) [inline Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:455]: Nice change. But it seems unrelated to this patch. Can it be moved to a separate patch?

**Review by smfr — APPROVED** [review]

File: Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp [hunk 2]
```diff
@@ -333,197 +314,94 @@ void Recorder::drawPattern(ImageBuffer& imageBuffer, const FloatRect& destRect,
...
-void Recorder::save(GraphicsContextState::Purpose purpose)
+void Recorder::updateStateForSave(GraphicsContextState::Purpose purpose)
 {
     ASSERT(purpose == GraphicsContextState::Purpose::SaveRestore);
-
     appendStateChangeItemIfNecessary();
     GraphicsContext::save(purpose);
-    recordSave();
     m_stateStack.append(m_stateStack.last());
 }
```
Comment (shallawa) [inline Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:323]: This function is not virtual and its name is `updateStateForSave()` and it is called from `RecorderImpl::save()` and `RemoteDisplayListRecorderProxy::save()`.  But it calls `GraphicsContext::save()`.

I think this naming is a little awkward and I do not think it will save much by removing this virtual call indirection if we keep this function virtual under the name `save()`

File: Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp [hunk 3]
```diff
@@ -333,197 +314,94 @@ void Recorder::drawPattern(ImageBuffer& imageBuffer, const FloatRect& destRect,
...
-        return;
+        return false;
 
     m_stateStack.removeLast();
-    recordRestore();
+    return true;
 }
 
-void Recorder::translate(float x, float y)
+bool Recorder::updateStateForTranslate(float x, float y)
```
Comment (shallawa) [inline Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:338]: I think this change is fine because `Recorder::translate()` does not call `GraphicsContext::translate()`

File: Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp [hunk 4]
```diff
@@ -333,197 +314,94 @@ void Recorder::drawPattern(ImageBuffer& imageBuffer, const FloatRect& destRect,
...
-    recordSetCTM(transform);
 }
 
 AffineTransform Recorder::getCTM(GraphicsContext::IncludeDeviceScale) const
 {
     return currentState().ctm;
 }
 
-void Recorder::beginTransparencyLayer(float opacity)
+void Recorder::updateStateForBeginTransparencyLayer(float opacity)
```
Comment (shallawa) [inline Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:381]: This is similar to save(). It's called from sub class `beginTransparencyLayer()` and it is going to call the base class `beginTransparencyLayer()`.

It looks like the `Recorder` implementation of `beginTransparencyLayer()`. It does not really look like an `updateStateForBeginTransparencyLayer()`

File: Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp [hunk 5]
```diff
@@ -333,197 +314,94 @@ void Recorder::drawPattern(ImageBuffer& imageBuffer, const FloatRect& destRect,
...
-
-void Recorder::fillRectWithRoundedHole(const FloatRect& rect, const FloatRoundedRect& roundedHoleRect, const Color& color)
-{
-    appendStateChangeItemIfNecessary();
-    recordFillRectWithRoundedHole(rect, roundedHoleRect, color);
-}
-
 void Recorder::fillPath(const Path& path)
 {
     appendStateChangeItemIfNecessary();
```
Comment (shallawa) [inline Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:407]: Maybe this function can have only one switch statement to direct the call to GraphicsContext calls:

```
    if (auto segment = path.singleSegment()) {
#if ENABLE(INLINE_PATH_DATA)
        if (auto line = path.singleDataLine())
            fillPathLine(*line);
        else if (auto arc = path.singleArc())
            fillPathArc(*arc);
        else if (auto closedArc = path.singleClosedArc())
            fillPathClosedArc(*closedArc);
        else if (auto curve = path.singleQuadCurve())
            fillPathQuadCurve(*curve);
        else if (auto curve = path.singleBezierCurve())
            fillPathBezierCurve(*curve);
        else
#endif
            fillPathSegment(*segment);
        return;
    }

    fillFullPath(path);
```

But I guess you plan to fix this in part 2.

File: Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h [hunk 6]
```diff
@@ -41,55 +41,69 @@ class RecorderImpl : public Recorder {
 
     bool isEmpty() const { return m_displayList.isEmpty(); }
 
+    void save(GraphicsContextState::Purpose) final;
```
Comment (shallawa) [inline Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:44]: Why these function became public?

Comment (kkinnunen-apple) [reply Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:44]: The eventual intention is that something, at least a test, is able to call the functions directly.
Putting them as private is violation of the intention of the API:

```
Ref<DisplayListRecorderImpl> m_context;
GraphicsContextGL* a = m_context.get();
a->drawPath(..); // works
m_context->drawPath(..); // should work too! It's the same function. This is a better function to call in cases it is possible. 
```

File: Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp (hunk 5 above)
Comment (kkinnunen-apple) [reply Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:407]: Yes. Eventually all these cases should be just removed. I don't think this *should* be an optimization the way it's written.

File: Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp (hunk 4 above)
Comment (kkinnunen-apple) [reply Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:381]: The intention is:
- Each drawing command that modifies state should call `Recorder::updateStateForABCD()`.
- Recoder is free to call whatever it needs to implement such state modifications to the GraphicsContext state tracking.

It is infact a GraphicsContext base class bug that the subclasses must call the vfunc base implementations to modify the state.

It would look odd that all state modification functions would call `updateStateForABCD()` but then save and `beginTransparencyLayer()` would call  `Recorder::save()` and `Recorder::beginTransparencyLayer()`. You can also note that the return types and parameters are different between `Recorder::ABCD()` and  `Recorder::updateStateForABCD()`. The intention is to not call the base class version, rather call the state update function that is implemented in whatever way is needed..

File: Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp (hunk 1 above)
Comment (kkinnunen-apple) [reply Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:455]: This patch kind of depends on this, and it's such a small change to get reviewed with latency. I'll go ahead and land as is, since I don't have that much time scheduled to work on this. Hope this is ok.
